### ISSUE TITLE
Maintain only one txt2tags version

### ISSUE THREAD

**Issue by jendrikseipp** [txt2tags/txt2tags#190]
[I originally posted this to the mailing list, but the message is probably held there.]

I think it is very unfortunate that there are two txt2tags versions in the repository at the moment, the original single-file txt2tags and the modular txt2tagslite. If we want to move the code forward, fix bugs and add features, we will have to decide on a single version. Maintaining two separate versions is just too much work.

Personally, I believe that the code complexity has grown too much for everything to be in one file. We will have to split up the code, make it modular and therefore easier to understand, change and extend. Txt2tagslite has done the first steps in the right direction, in my opinion. If we decide to stick with it we can use it to move even more code from the main file into the target modules.

Having only one file has little to no benefit, whereas organized code in submodules can greatly simplify the development. Therefore, I vote to only keep txt2tagslite in the codebase (and rename it to txt2tags). 

What are your thoughts?

**Comment by farvardin** [comment]
I think html support is a very important feature. Therefore, even if 
other targets are kept in a separate extension, the main txt2tags tool 
should contain at least a HTML export.

To keep extensions or libraries outside the core for easing the 
development is OK, but a way to "concatenate" everything in one file 
only would be easier for users (and for me too).

I've never managed to make txt2tagslite work for me.

Also, I believe txt2tags is more than just a python script (which I 
still use 3-4 times a week): there are a php implementations (so for 
html export only), which I use every day since it is powering my 
personal wiki and my websites, and also a javascript one. More 
implementations (ruby, c etc) would mean more users of the txt2tags markup.

**Comment by jendrikseipp** [comment]
Let me rephrase my proposal: there has to be a repository with a *single*, *reference* txt2tags implementation. 

All other implementations (in other languages, in a single python file) can live in separate repos. We have to make it clear how the language is supposed to behave by selecting *one* official reference implementation.

By cleaning up the repo and selecting one reference python implementation, we make it easier for us to maintain the project and for others to contribute to it.

Or more pragmatically: one of the worst things in programming is code duplication and currently there are about 6000 lines of duplicated code in txt2tagslite. We have to choose one implementation or otherwise the project can't move forward.

**Comment by smartmic** [comment]
Very good idea ! 
I would really like to see one repo as a single reference. As already posted to the old Google mailing list, I have prepared a update for the Lout target which was some effort and I would be happy to share. 
Implemented in txt2tags, I learned during testing, that there is a second implementation txt2tagslite which I was not aware of at all. So I have stalled final implementation and look forward to reference Github repo now. 
In my opinion, having one single txt2tags file is a neat concept and was one reason for me to try out txt2tags some time ago.

**Comment by aureliojargas** [comment]
Maybe have the best of both worlds is feasible?

* The original source code in GitHub is the current txt2tagslite: the code is spread across several files.
* The released source code is a single Python file, generated from a script that joins it all together.

So the development occurs in separated files, which are all joined in a single file to make an official release.

That's what I do in my [Funções ZZ project](https://github.com/funcoeszz/funcoeszz). It's a shell script software, maybe it's easier to do in that language. But I guess a smart script could do the packaging for txt2tags.

**Comment by farvardin** [comment]
yes, I think it's the way to go.
I hope it's not too difficult to do this in python as well. Joining all the
files together means modifying the source to remove the includes.

2015-06-09 13:16 GMT+02:00 Aurelio Jargas <notifications@github.com>:

**Comment by jendrikseipp** [comment]
In my opinion, there could be a separate txt2tags-single-file project that
makes releases independently from the main project. Whether this is done
solely by a script or if some tuning by hand is required, is not really
important then.

**Comment by fgallaire** [comment]
Hello, so much things to tell about that.

1) About the duplication
There's no real duplication, because txt2tagslite was at first an experimentation and never officially released. However, it's stable and pass all tests since a long time.

2) About the one file concept
For some reasons, some users love the historical "one file concept" (easy to use and deploy) and devs hate it (no need to explain why a 10 000 LoC file is not good engineering). But as @smartmic said, it's an easy way to start a hack !

3) About the future
I hope the one file txt2tags will be released for a last time as txt2tags 2.7, and will not be supported anymore. Txt2tagslite will be released as txt2tagslite 2.7 and will change of name after that #192.

**Comment by jendrikseipp** [comment]
If we want to keep both implementations around until txt2tags 2.7, then let's release it soon, please.

**Comment by farvardin** [comment]
On 29/06/2015 05:01, Florent Gallaire wrote:
>
> Hello, so much things to tell about that.
>
> 1) About the duplication
> There's no real duplication, because txt2tagslite was at first an 
> experimentation and never officially released. However, it's stable 
> and pass all tests since a long time.
>
> 2) About the one file concept
> For some reasons, some users love the historical "one file concept" 
> (easy to use and deploy) and devs hate it (no need to explain why a 10 
> 000 LoC file is not good engineering). But as @smartmic 
> <https://github.com/smartmic> said, it's an easy way to start a hack !
>
> 3) About the future
> I hope the one file txt2tags will be released for a last time as 
> txt2tags 2.7, and will not be supported anymore. Txt2tagslite will be 
> released as txt2tagslite 2.7 and will change of name after that #192 
> <https://github.com/txt2tags/txt2tags/issues/192>.
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/txt2tags/txt2tags/issues/190#issuecomment-116399824>.
>

I don't care if the "just one file" concept is no longer kept for the 
development's sake, probably having several files is easier for 
developpers, but several files doesn't justify at all for the final 
release, especially when it's only 300 kb.

**Comment by jendrikseipp** [comment]
I just found out about an interesting Python feature: zip applications. You can package python files in a zip file (called e.g. myzipfile.pyz) and run it with `python myzipfile.pyz`. This feature is exactly what we need: we can give users a single executable file and use multiple files for development.

The feature has been present since Python 2.6, but is probably getting more traction now that Python 3.5 offers a module for creating zip apps. They can, however, also be built manually without this new module.

https://docs.python.org/dev/library/zipapp.html#the-python-zip-application-archive-format

In light of this, I propose to only keep the current txt2tagslite implementation as soon as possible. We should not wait until a new version is released, because we need to fix bugs and tests now. Can I prepare a pull request, Aurélio? Everyone else should check that txt2tagslite really is up to par with txt2tags.

**Comment by aureliojargas** [comment]
The current maintainers (@jendrikseipp @farvardin @fgallaire) should decide what's best for the project now. I'm here just listening and giving random advices :)

**Comment by fgallaire** [comment]
@jendrikseipp we are very defensive about the Python version number (@aureliojargas legacy). For our historical users, we have to release txt2tags 2.7 in the old school way. We will fix both txt2tags and txt2tagslite until the release, then @farvardin will probably maintain the old one file txt2tgas, and me try to improve the new txt2tagslite way, where your new ideas could be feasible.

**Comment by jendrikseipp** [comment]
With the zip application we can support Python 2.6 users and that Python version has been released 7 years ago (2008). I don't think that we need to support any older Python installations, especially since that makes it much harder to support Python 3.x, and Python 3.0 was also released in 2008... I can fully understand your reservations, but trying to maintain both txt2tags and txt2tagslite while fixing errors and hopefully also the failing tests (of which there are many) will be a lot of work and prolong the release of version 2.7 significantly.

**Comment by jendrikseipp** [comment]
Another solution for removing the code duplication is to move txt2tagslite to a separate repo, fix txt2tags, release txt2tags 2.7, and then apply the current diff between the two versions to make txt2tags more modular.

So we have two options: either we remove txt2tags or we remove txt2tagslite. Personally, I would prefer to remove txt2tags since modularity should be the ultimate goal and integrating the txt2tagslite changes later might be more work. However, I wouldn't mind removing txt2tagslite either. The important thing is to remove this horrible code duplication. Which option do you guys prefer?

**Comment by fgallaire** [comment]
@jendrikseipp it's not that simple. We stay like this until the release.

**Comment by jendrikseipp** [comment]
Instead of duplicating the code in one repository, txt2tagslite should have been a fork from the beginning. I can understand that forking with svn might have been difficult, but now that we use git, it has become very simple. After txt2tags 2.7 is released we can merge the fork back into the main repository. Can you explain why you think it's not that simple?

**Comment by jendrikseipp** [comment]
I think this issue has been resolved by the split into the v2 and v3 branches. By the way, I don't think anymore that splitting txt2tags into multiple files is required for more modular code. A single file can hold modular code as well.
